perception/autoware_tensorrt_bevdet/src/preprocess_plugin.cu

```diff
@@ -0,0 +1,347 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "autoware/tensorrt_bevdet/common.hpp"
+#include "autoware/tensorrt_bevdet/preprocess_plugin.hpp"
+
+#include <cuda_fp16.h>
+#include <cuda_runtime_api.h>
+
+#include <iostream>
+
+// kernel for GPU
+
+template <typename T>
+__global__ void preprocess_kernel(
+  const uint8_t * src_dev, T * dst_dev, int src_row_step, int dst_row_step, int src_img_step,
+  int dst_img_step, int src_h, int src_w, float radio_h, float radio_w, float offset_h,
+  float offset_w, const float * mean, const float * std, int dst_h, int dst_w, int n)
+{
+  int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= dst_h * dst_w * n) return;
+
+  int i = (idx / n) / dst_w;
+  int j = (idx / n) % dst_w;
+  int k = idx % n;
+
+  int pX = (int)roundf((i / radio_h) + offset_h);
+  int pY = (int)roundf((j / radio_w) + offset_w);
+
+  if (pX < src_h && pX >= 0 && pY < src_w && pY >= 0) {
+    int s1 = k * src_img_step + 0 * src_img_step / 3 + pX * src_row_step + pY;
+    int s2 = k * src_img_step + 1 * src_img_step / 3 + pX * src_row_step + pY;
+    int s3 = k * src_img_step + 2 * src_img_step / 3 + pX * src_row_step + pY;
+
+    int d1 = k * dst_img_step + 0 * dst_img_step / 3 + i * dst_row_step + j;
+    int d2 = k * dst_img_step + 1 * dst_img_step / 3 + i * dst_row_step + j;
+    int d3 = k * dst_img_step + 2 * dst_img_step / 3 + i * dst_row_step + j;
+
+    dst_dev[d1] = (static_cast<T>(src_dev[s1]) - static_cast<T>(mean[0])) / static_cast<T>(std[0]);
+    dst_dev[d2] = (static_cast<T>(src_dev[s2]) - static_cast<T>(mean[1])) / static_cast<T>(std[1]);
+    dst_dev[d3] = (static_cast<T>(src_dev[s3]) - static_cast<T>(mean[2])) / static_cast<T>(std[2]);
+  }
+}
+
+namespace nvinfer1
+{
+// class PreprocessPlugin
+PreprocessPlugin::PreprocessPlugin(
+  const std::string & name, int crop_h, int crop_w, float resize_radio)
+: name_(name)
+{
+  m_.crop_h = crop_h;
+  m_.crop_w = crop_w;
+  m_.resize_radio = resize_radio;
+}
+
+PreprocessPlugin::PreprocessPlugin(const std::string & name, const void * buffer, size_t length)
+: name_(name)
+{
+  memcpy(&m_, buffer, sizeof(m_));
+}
+
+PreprocessPlugin::~PreprocessPlugin()
+{
+}
+
+IPluginV2DynamicExt * PreprocessPlugin::clone() const noexcept
+{
+  auto p = new PreprocessPlugin(name_, &m_, sizeof(m_));
+  p->setPluginNamespace(namespace_.c_str());
+  return p;
+}
+
+int32_t PreprocessPlugin::getNbOutputs() const noexcept
+{
+  return 1;
+}
+
+DataType PreprocessPlugin::getOutputDataType(
+  int32_t index, DataType const * inputTypes, int32_t nbInputs) const noexcept
+{
+  // return DataType::kHALF;
+  return DataType::kFLOAT;
+}
+
+DimsExprs PreprocessPlugin::getOutputDimensions(
+  int32_t outputIndex, const DimsExprs * inputs, int32_t nbInputs,
+  IExprBuilder & exprBuilder) noexcept
+{
+  int input_h = inputs[0].d[2]->getConstantValue();
+  int input_w = inputs[0].d[3]->getConstantValue() * 4;
```

Comment:
The calculation `input_w = inputs[0].d[3]->getConstantValue() * 4;` assumes the input image has 4 channels (e.g., RGBA). This should be made more explicit, either through a comment or by adding a parameter to handle different channel counts.  If it's always 4 channels, the comment should clearly state that.